src/librustc_trans/trans/mir/block.rs

```diff
@@ -46,8 +47,23 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
                 build::CondBr(bcx, cond.immediate(), lltrue, llfalse, DebugLoc::None);
             }
 
-            mir::Terminator::Switch { .. } => {
-                unimplemented!()
+            mir::Terminator::Switch { ref discr, ref adt_def, ref targets } => {
+                let adt_ty = bcx.tcx().lookup_item_type(adt_def.did).ty;
+                let represented_ty = adt::represent_type(bcx.ccx(), adt_ty).clone();
+
+                let discr_lvalue = self.trans_lvalue(bcx, discr);
+                let discr = adt::trans_get_discr(bcx, &*represented_ty, discr_lvalue.llval, None);
+
+                //Use the diverge block for the default case since all enum variants have targets
```

Comment:
For exhaustive matches, the default arm of the switch statement should go to a block that only contains an `unreachable` instruction. LLVM uses that for optimizations.